vsphere/resource_vsphere_vmfs_datastore.go

```diff
@@ -0,0 +1,324 @@
+WARNING: Dangling resource!
+After creating the datastore, there was an error fetching its properties:
+%s
+Additionally, there was an error removing the created datastore:
+%s
+You will need to remove this datastore manually before trying again.
+`
+
+// formatUpdateInconsistentState defines the verbose error when the setting
+// state failed in the middle of an update opeartion. This is an error that
+// will require repair of the state before TF can continue.
+const formatUpdateInconsistentState = `
+WARNING: Inconsistent state!
+Terraform was able to add disk %s, but could not save the update to the state.
+The error was:
+%s
+This is more than likely a bug. Please report it at:
+https://github.com/terraform-providers/terraform-provider-vsphere/issues
+You will also need to repair your state before trying again.
+`
+
+func resourceVSphereVmfsDatastore() *schema.Resource {
+	s := map[string]*schema.Schema{
+		"name": &schema.Schema{
+			Type:        schema.TypeString,
+			Description: "The name of the datastore.",
+			Required:    true,
+			ForceNew:    true,
+		},
+		"host_system_id": &schema.Schema{
+			Type:        schema.TypeString,
+			Description: "The managed object ID of the host to set up the datastore on.",
+			Required:    true,
+		},
+		"disks": &schema.Schema{
+			Type:        schema.TypeList,
+			Description: "The disks to add to the datastore.",
+			Required:    true,
+			MinItems:    1,
+			Elem:        &schema.Schema{Type: schema.TypeString},
+		},
+	}
+	mergeSchema(s, schemaDatastoreSummary())
+	return &schema.Resource{
+		Create: resourceVSphereVmfsDatastoreCreate,
+		Read:   resourceVSphereVmfsDatastoreRead,
+		Update: resourceVSphereVmfsDatastoreUpdate,
+		Delete: resourceVSphereVmfsDatastoreDelete,
+		Schema: s,
+	}
+}
+
+func resourceVSphereVmfsDatastoreCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+	hsID := d.Get("host_system_id").(string)
+	dss, err := hostDatastoreSystemFromHostSystemID(client, hsID)
+	if err != nil {
+		return fmt.Errorf("error loading host datastore system: %s", err)
+	}
+
+	// To ensure the datastore is fully created with all the disks that we want
+	// to add to it, first we add the initial disk, then we expand the disk with
+	// the rest of the extents.
+	disks := d.Get("disks").([]interface{})
+	disk := disks[0].(string)
+	spec, err := diskSpecForCreate(dss, disk)
+	if err != nil {
+		return err
+	}
+	spec.Vmfs.VolumeName = d.Get("name").(string)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	ds, err := dss.CreateVmfsDatastore(ctx, *spec)
+	if err != nil {
+		return fmt.Errorf("error creating datastore with disk %s: %s", disk, err)
+	}
+
+	// Now add any remaining disks.
+	for _, disk := range disks[1:] {
+		spec, err := diskSpecForExtend(dss, ds, disk.(string))
+		if err != nil {
+			// We have to destroy the created datastore here.
+			if remErr := removeDatastore(dss, ds); remErr != nil {
```

Comment:
as above, can we track this with `d.Partial` to allow us to re-attempt this in the a future plan?

Reply:
I think I'd like to keep the `Create` operation atomic - creating a datastore and adding extents is not a ridiculously expensive operation, and if a user is expecting a datastore to be created with a certain number of extents and one of the extents is no good for the datastore, I think it makes more sense to not have a datastore at all versus a partially created one.